kratos/utilities/parallel_distance_calculator.cpp

```diff
@@ -0,0 +1,630 @@
+
+    CalculateExactDistancesOnDividedElements(rModelPart, rDistanceVar, rAreaVar, MaxDistance, Options);
+
+    ExtendDistancesByLayer(rModelPart, rDistanceVar, rAreaVar, MaxLevels, MaxDistance);
+
+    AssignDistanceSign(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    KRATOS_CATCH("")
+}
+
+template<unsigned int TDim>
+void ParallelDistanceCalculator<TDim>::CalculateInterfacePreservingDistances(
+    ModelPart& rModelPart,
+    const Variable<double>& rDistanceVar,
+    const Variable<double>& rAreaVar,
+    const unsigned int MaxLevels,
+    const double MaxDistance)
+{
+    KRATOS_TRY
+
+    Check(rModelPart, rDistanceVar, rAreaVar);
+
+    ResetVariables(rModelPart,rDistanceVar, MaxDistance);
+
+    AbsDistancesOnDividedElements(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    ExtendDistancesByLayer(rModelPart, rDistanceVar, rAreaVar, MaxLevels, MaxDistance);
+
+    AssignDistanceSign(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    KRATOS_CATCH("")
+}
+
+template<unsigned int TDim>
+void ParallelDistanceCalculator<TDim>::CalculateDistancesLagrangianSurface(
+    ModelPart& rModelPart,
+    const Variable<double>& rDistanceVar,
+    const Variable<double>& rAreaVar,
+    const unsigned int MaxLevels,
+    const double MaxDistance)
+{
+    KRATOS_TRY
+
+    //check that variables needed are in the model part
+    const bool is_distributed = rModelPart.IsDistributed();
+    KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(rDistanceVar)) << "Distance variable is not in the model part" << std::endl;
+    KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(rAreaVar)) << "Area Variable is not in the model part" << std::endl;
+    if(is_distributed)
+        KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(PARTITION_INDEX)) << "PARTITION_INDEX variable is not in the model part" << std::endl;
+
+    // set to zero the distance
+    block_for_each(rModelPart.Nodes(), [&](NodeType& rNode){
+        double& r_area = rNode.FastGetSolutionStepValue(rAreaVar);
+        double& r_distance = rNode.FastGetSolutionStepValue(rDistanceVar);
+        rNode.GetValue(rDistanceVar) = r_distance;
+        if (rNode.IsNot(VISITED)) {
+            r_area = 0.0;
+            r_distance = 0.0;
+        } else {
+            r_area = 1.0;
+        }
+    });
+
+    // Set the TLS container
+    array_1d<double,TDim+1> visited, N;
+    BoundedMatrix <double, TDim+1,TDim> DN_DX;
+    typedef std::tuple<array_1d<double,TDim+1>, array_1d<double,TDim+1>, BoundedMatrix<double, TDim+1, TDim>> TLSType;
+    TLSType tls_container = std::make_tuple(visited, N, DN_DX);
+
+    // Extend the distances layer by layer up to a maximum level of layers
+    for(unsigned int level=0; level<MaxLevels; level++)
+    {
+        //loop on active elements and advance the distance computation
+        block_for_each(rModelPart.Elements(), tls_container, [&](Element& rElement, TLSType& rTLSContainer){
```

Comment:
Is it possible to do the following?

```suggestion
        block_for_each(rModelPart.Elements(), TLSType(), [&](Element& rElement, TLSType& rTLSContainer){
```

So that we can remove `tls_container`

Reply:
All the variables in `TLSType` are stack variables, do we have a gain in using `TLS` here?

Reply:
You allocate once per thread rather than once per element.